File: Tree10/WordSearch2.py
# https://leetcode.com/problems/word-search-ii/description/
class WordSearch2:

    # Time : 4^r*c, Space : O(2(r*c))
	def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:

        result = []
        charSet = {word[0] for word in words}
       
        row = len(board)
        col = len(board[0])
        visited = [([0] * col) for i in range(row)]

        for r in range (row):
            for c in range (col):
                if board[r][c] in charSet:
                    self.dfs(r, c, board, words, result, "", visited)
        
        return result


    def dfs(self, row, col, board, words, result, word, visited):
        if row < 0 or col < 0 or row >= len(board) or col >= len(board[0]) or visited[row][col] == 1: return 

        word += board[row][col]
        if word in words and word not in result: 
            result.append(word)
        
        visited[row][col] = 1
        self.dfs(row - 1, col, board, words, result, word, visited)
        self.dfs(row, col - 1, board, words, result, word, visited)
        self.dfs(row + 1, col, board, words, result, word, visited)
        self.dfs(row, col + 1, board, words, result, word, visited)
        visited[row][col] = 0


# --------------------------------------------------------------------------------------------------

# create a trie on words 
class TrieNode:

    def __init__(self):
        self.childrens = {}
        self.isEnd = False

    def addWord(self, word):
        current = self
        for c in word:
            if c not in current.childrens:
                current.childrens[c] = TrieNode()
        
            current = current.childrens[c]
        current.isEnd = True


class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        root = TrieNode()
        for w in words: root.addWord(w)

        rows, cols = len(board), len(board[0])
        result, visited = set(), set()
    
        # peform dfs on all charin board and if its in trie
        def dfs(row, col, node, word):
            if (row < 0 or col < 0 or row >= rows or col >= cols or 
                (row, col) in visited or  board[row][col] not in node.childrens): 
                return                 
            
            # update status
            visited.add((row, col))
            node = node.childrens[board[row][col]]
            word += board[row][col]
            
            # If we find a word
            if node.isEnd:
                result.add(word)

            # Explore all 4 directions
            dfs(row - 1, col, node, word)  # Up
            dfs(row + 1, col, node, word)  # Down
            dfs(row, col - 1, node, word)  # Left
            dfs(row, col + 1, node, word)  # Right

            # Backtrack
            visited.remove((row, col))

        #  for each char 
        for r in range(rows):
            for c in range(cols):
                dfs(r, c, root, '')

        return list(result)

File: Graph14/WordSearch2.java
public class WordSearch2{

	public List<String> findWords(char[][] board, String[] words) {
        
        List<String> result = new ArrayList<>();
        for(String word: words){
            if(searchWord(word, board)){
                result.add(word);
            }
        }

        return result;
    }

    public boolean searchWord(String word, char[][] board){
        char[] charArr = word.toCharArray();
        int[][] visited = new int[board.length][board[0].length];

        for(int i=0; i<board.length; i++){
            for(int j=0; j<board[i].length; j++){
                if(board[i][j] == charArr[0]){
                    if(dfs(0, i, j, board,charArr, visited)){
                        return true;
                    }
                }
            }
        }

        return false;
    }


    public boolean dfs(int i, int r, int c, char[][] board,char[] charArr, int[][] visited){
        if(i>=charArr.length) return true;
        if(r < 0 || c < 0 || r>=board.length || c>=board[0].length || visited[r][c] == 1 || board[r][c] != charArr[i]) return false;

        visited[r][c] = 1;

        if(dfs(i+1, r+1, c, board, charArr, visited)) return true;
        if(dfs(i+1, r-1, c, board, charArr, visited)) return true;
        if(dfs(i+1, r, c+1, board, charArr, visited)) return true;
        if(dfs(i+1, r, c-1, board, charArr, visited)) return true;

        visited[r][c] = 0;

        return false;
    }
}


    // -------------------------------------------------------- Trie ----------------------------------------------------------------

class Trie{
    Trie[] child;
    boolean isEnd;

    Trie(){
        child = new Trie[26];
        isEnd = false;
    }

    public void insertWord(String word){
        Trie node = this;
        for(char ch: word.toCharArray()){
            if(node.child[ch - 'a'] == null){
                node.child[ch - 'a'] = new Trie();
            }
            node = node.child[ch - 'a'];
        }
        node.isEnd = true;
    }

    public Trie get(char ch){
        return child[ch-'a'];
    }
}


public class WordSearch2 {
    public List<String> findWords(char[][] board, String[] words) {
        List<String> result = new ArrayList<>();
        Trie trie = new Trie();

        for(String word: words) {
            trie.insertWord(word);
        }

        int[][] visited = new int[board.length][board[0].length];

        for(int r=0; r<board.length; r++){
            for(int c=0; c<board[r].length; c++){
                dfs(r, c, board, visited, "", trie, result);
            }
        }

        return result;
    }

    public void dfs(int r, int c, char[][] board, int[][] visited, String word, Trie trie, List<String> result){
        if(r<0 || c<0 || r>= board.length || c>= board[0].length || visited[r][c] == 1 || trie.get(board[r][c]) == null) return;

        visited[r][c] = 1;
        trie = trie.get(board[r][c]);
        word+=board[r][c];
        
        if(trie.isEnd){
            result.add(word);
            trie.isEnd = false;
        }

        dfs(r+1, c, board, visited, word, trie, result);
        dfs(r-1, c, board, visited, word, trie, result);
        dfs(r, c+1, board, visited, word, trie, result);
        dfs(r, c-1, board, visited, word, trie, result);

        visited[r][c] = 0;
    }
}